File: cucumber-expressions/go/ast.go
package cucumberexpressions

import (
	"strings"
	"unicode"
)

const escapeCharacter rune = '\\'
const alternationCharacter rune = '/'
const beginParameterCharacter rune = '{'
const endParameterCharacter rune = '}'
const beginOptionalCharacter rune = '('
const endOptionalCharacter rune = ')'

type nodeType string

const (
	textNode        nodeType = "TEXT_NODE"
	optionalNode    nodeType = "OPTIONAL_NODE"
	alternationNode nodeType = "ALTERNATION_NODE"
	alternativeNode nodeType = "ALTERNATIVE_NODE"
	parameterNode   nodeType = "PARAMETER_NODE"
	expressionNode  nodeType = "EXPRESSION_NODE"
)

type node struct {
	NodeType nodeType `json:"type"`
	Start    int      `json:"start"`
	End      int      `json:"end"`
	Token    string   `json:"token"`
	Nodes    []node   `json:"nodes"`
}

func (node node) text() string {
	builder := strings.Builder{}
	builder.WriteString(node.Token)

	if node.Nodes == nil {
		return builder.String()
	}

	for _, c := range node.Nodes {
		builder.WriteString(c.text())
	}
	return builder.String()
}

type tokenType string

const (
	startOfLine    tokenType = "START_OF_LINE"
	endOfLine      tokenType = "END_OF_LINE"
	whiteSpace     tokenType = "WHITE_SPACE"
	beginOptional  tokenType = "BEGIN_OPTIONAL"
	endOptional    tokenType = "END_OPTIONAL"
	beginParameter tokenType = "BEGIN_PARAMETER"
	endParameter   tokenType = "END_PARAMETER"
	alternation    tokenType = "ALTERNATION"
	text           tokenType = "TEXT"
)

type token struct {
	Text      string    `json:"text"`
	TokenType tokenType `json:"type"`
	Start     int       `json:"start"`
	End       int       `json:"end"`
}

var nullNode = node{textNode, -1, -1, "", nil}

func isEscapeCharacter(r rune) bool {
	return r == escapeCharacter
}

func canEscape(r rune) bool {
	if unicode.Is(unicode.White_Space, r) {
		return true
	}
	switch r {
	case escapeCharacter:
		return true
	case alternationCharacter:
		return true
	case beginParameterCharacter:
		return true
	case endParameterCharacter:
		return true
	case beginOptionalCharacter:
		return true
	case endOptionalCharacter:
		return true
	}
	return false
}

func typeOf(r rune) (tokenType, error) {
	if unicode.Is(unicode.White_Space, r) {
		return whiteSpace, nil
	}
	switch r {
	case alternationCharacter:
		return alternation, nil
	case beginParameterCharacter:
		return beginParameter, nil
	case endParameterCharacter:
		return endParameter, nil
	case beginOptionalCharacter:
		return beginOptional, nil
	case endOptionalCharacter:
		return endOptional, nil
	}
	return text, nil
}

func symbol(tokenType tokenType) string {
	switch tokenType {
	case beginOptional:
		return string(beginOptionalCharacter)
	case endOptional:
		return string(endOptionalCharacter)
	case beginParameter:
		return string(beginParameterCharacter)
	case endParameter:
		return string(endParameterCharacter)
	case alternation:
		return string(alternationCharacter)
	}

	return ""
}

func purpose(tokenType tokenType) string {
	switch tokenType {
	case beginOptional:
		return "optional text"
	case endOptional:
		return "optional text"
	case beginParameter:
		return "a parameter"
	case endParameter:
		return "optional text"
	case alternation:
		return "alternation"
	}

	return ""
}

File: cucumber-expressions/javascript/src/Ast.ts
const escapeCharacter = '\\'
const alternationCharacter = '/'
const beginParameterCharacter = '{'
const endParameterCharacter = '}'
const beginOptionalCharacter = '('
const endOptionalCharacter = ')'

export function symbolOf(token: TokenType): string {
  switch (token) {
    case TokenType.beginOptional:
      return beginOptionalCharacter
    case TokenType.endOptional:
      return endOptionalCharacter
    case TokenType.beginParameter:
      return beginParameterCharacter
    case TokenType.endParameter:
      return endParameterCharacter
    case TokenType.alternation:
      return alternationCharacter
  }
  return ''
}

export function purposeOf(token: TokenType): string {
  switch (token) {
    case TokenType.beginOptional:
    case TokenType.endOptional:
      return 'optional text'
    case TokenType.beginParameter:
    case TokenType.endParameter:
      return 'a parameter'
    case TokenType.alternation:
      return 'alternation'
  }
  return ''
}

export interface Located {
  readonly start: number
  readonly end: number
}

export class Node {
  readonly type: NodeType
  readonly nodes?: readonly Node[] | undefined
  readonly token?: string | undefined
  readonly start: number
  readonly end: number

  constructor(
    type: NodeType,
    nodes: readonly Node[] = undefined,
    token: string = undefined,
    start: number,
    end: number
  ) {
    if (nodes === undefined && token === undefined) {
      throw new Error('Either nodes or token must be defined')
    }
    if (nodes === null || token === null) {
      throw new Error('Either nodes or token may not be null')
    }
    this.type = type
    this.nodes = nodes
    this.token = token
    this.start = start
    this.end = end
  }

  text(): string {
    if (this.nodes) {
      return this.nodes.map((value) => value.text()).join('')
    }
    return this.token
  }
}

export enum NodeType {
  text = 'TEXT_NODE',
  optional = 'OPTIONAL_NODE',
  alternation = 'ALTERNATION_NODE',
  alternative = 'ALTERNATIVE_NODE',
  parameter = 'PARAMETER_NODE',
  expression = 'EXPRESSION_NODE',
}

export class Token {
  readonly type: TokenType
  readonly text: string
  readonly start: number
  readonly end: number

  constructor(type: TokenType, text: string, start: number, end: number) {
    this.type = type
    this.text = text
    this.start = start
    this.end = end
  }

  static isEscapeCharacter(codePoint: string): boolean {
    return codePoint == escapeCharacter
  }

  static canEscape(codePoint: string): boolean {
    if (codePoint == ' ') {
      // TODO: Unicode whitespace?
      return true
    }
    switch (codePoint) {
      case escapeCharacter:
        return true
      case alternationCharacter:
        return true
      case beginParameterCharacter:
        return true
      case endParameterCharacter:
        return true
      case beginOptionalCharacter:
        return true
      case endOptionalCharacter:
        return true
    }
    return false
  }

  static typeOf(codePoint: string): TokenType {
    if (codePoint == ' ') {
      // TODO: Unicode whitespace?
      return TokenType.whiteSpace
    }
    switch (codePoint) {
      case alternationCharacter:
        return TokenType.alternation
      case beginParameterCharacter:
        return TokenType.beginParameter
      case endParameterCharacter:
        return TokenType.endParameter
      case beginOptionalCharacter:
        return TokenType.beginOptional
      case endOptionalCharacter:
        return TokenType.endOptional
    }
    return TokenType.text
  }
}

export enum TokenType {
  startOfLine = 'START_OF_LINE',
  endOfLine = 'END_OF_LINE',
  whiteSpace = 'WHITE_SPACE',
  beginOptional = 'BEGIN_OPTIONAL',
  endOptional = 'END_OPTIONAL',
  beginParameter = 'BEGIN_PARAMETER',
  endParameter = 'END_PARAMETER',
  alternation = 'ALTERNATION',
  text = 'TEXT',
}
